### ISSUE TITLE
PID 1: support handling loop devices without races

### ISSUE THREAD

**Issue by DemiMarie** [systemd/systemd#19150]
**Is your feature request related to a problem? Please describe.**
The Linux® kernel does not provide a race-free interface for handling loop devices.  Immediately after requesting a loop device from `/dev/loop-control`, unclean process termination will result in the loop device being leaked.  This window of time persists until `LO_FLAGS_AUTOCLEAR` is set on the device.

**Describe the solution you'd like**
As the global init process, systemd is in the unique position of running for as long as the system remains up.  Therefore, it is immune from the aforementioned race condition.  This can be exposed as a D-Bus API that returns both a file descriptor for the loop device (which already has `LO_FLAGS_AUTOCLEAR` set) as well as the device name.

**Describe alternatives you've considered**
A loop file system, loopfs, has been proposed for inclusion in Linux.  However, it is not yet part of Linux.

**The systemd version you checked that didn't have the feature you are asking for**
systemd 246

**Comment by marmarek** [comment]
I believe this very scenario is solved the by recent introduction of `LOOP_CONFIGURE` ioctl (atomic set_fd + set_status). But there may be other scenarios where such feature would be useful, as the kernel API for loop devices is awful and full of races.

**Comment by poettering** [comment]
> I believe this very scenario is solved by the recent introduction of `LOOP_CONFIGURE` ioctl (atomic set_fd + set_status). But there may be other scenarios where such feature would be useful, as the kernel API for loop devices is awful and full of races.

It's not enough unfortunately: for LO_FLAGS_AUTOCLEAR to become effective you actually need some action that makes it unused to happen (for example unmount it or so). But if you just allocate a loopback device, and then immediately die for some reason even with LO_FLAGS_AUTOCLEAR set the device is not going to be detached, because there never was a umount operation or equivalent taking place.  It's a racy mess hence.

I am not convinced this is something for userspace to work around though. I am pretty sure the kernel should fix this on their own. my idea would be: whenever a loopback device is attached, make that fd that did that special and consider it a reference that keeps the device attached and allocated as long as it is open, and then make the LO_FLAGS_AUTOCLEAR logic level-triggered instead of edge-triggered. This way there would be two kinds of fds to a loopback device: the regular kind which won#t keep the device attached, and the special kind which will and which you only get if your process is the one that attached the device. 

If so things would be robust: if you die too early, the special fd is dropped and the loopback device cleaned up.

**Comment by DemiMarie** [comment]
> > I believe this very scenario is solved by the recent introduction of `LOOP_CONFIGURE` ioctl (atomic set_fd + set_status). But there may be other scenarios where such feature would be useful, as the kernel API for loop devices is awful and full of races.
> 
> It's not enough unfortunately: for LO_FLAGS_AUTOCLEAR to become effective you actually need some action that makes it unused to happen (for example unmount it or so). But if you just allocate a loopback device, and then immediately die for some reason even with LO_FLAGS_AUTOCLEAR set the device is not going to be detached, because there never was a umount operation or equivalent taking place. It's a racy mess hence.

Interesting (and unfortunate).  I thought `LO_FLAGS_AUTOCLEAR` meant that a device would automatically be deallocated if there are no references to it.

> I am not convinced this is something for userspace to work around though. I am pretty sure the kernel should fix this on their own. my idea would be: whenever a loopback device is attached, make that fd that did that special and consider it a reference that keeps the device attached and allocated as long as it is open, and then make the LO_FLAGS_AUTOCLEAR logic level-triggered instead of edge-triggered. This way there would be two kinds of fds to a loopback device: the regular kind which won#t keep the device attached, and the special kind which will and which you only get if your process is the one that attached the device.
> 
> If so things would be robust: if you die too early, the special fd is dropped and the loopback device cleaned up.

The problem is `losetup(8)`: it is used by all sorts of scripts, but that makes it unsafe to close a loop device we did not open ourselves.  systemd could avoid this problem by providing a custom, stateful version of `losetup(8)` that returned `/dev/loop/GUID` instead.  GUIDs would be formed by encrypting a counter and so would never be reused.

**Comment by marmarek** [comment]
>  I thought `LO_FLAGS_AUTOCLEAR` meant that a device would automatically be deallocated if there are no references to it.

You are correct, closing a loop FD with `LO_FLAGS_AUTOCLEAR` set, that isn't referenced anywhere else (like mounted), will release it.

<details>
<summary>Test case</summary>

```
$ cat testloop.c 
#include <linux/loop.h>
#include <stdio.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <fcntl.h>

int main() {
    int i;
    struct loop_config lc = {
        .fd = 0,
        .info.lo_flags = LO_FLAGS_AUTOCLEAR,
    };
    char name[16];

    for (i=0; i<5; i++) {
        int fd = open("/dev/loop-control", O_RDWR);
        int loop_num = ioctl(fd, LOOP_CTL_GET_FREE, 0);
        close(fd);
        printf("%d\n", loop_num);
        sprintf(name, "/dev/loop%d", loop_num);
        int loop_fd = open(name, O_RDWR);
        if (ioctl(loop_fd, LOOP_CONFIGURE, &lc) == -1)
            perror("LOOP_CONFIGURE");
        close(loop_fd);
    }
    return 0;
}
$ sudo ./testloop < testfile.img 
0
0
0
0
1
$ sudo losetup -a
$
```
As you can see all are released automatically.
